### PR TITLE
STM32 Logger Firmware

### PR DESCRIPTION
**Pull request by JasonBrave**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+*.elf
```
Comment (Joe-Joe-Joe-Joe) [inline .gitignore:1]: I recommend using the following gitignore config from processor:
```*Backup*
STM32Cube/Debug/*
STM32Cube/Release/*
**/.metadata
**/.settings
**/.vscode```

File: COPYING
Comment (Joe-Joe-Joe-Joe) [inline COPYING:1]: What is this? Is this a build artifact?

File: Core/Src/error.c [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#include "error.h"
+#include "platform.h"
+
+static cansw_logger_error_t last_error;
+static bool any_errors_happened = false;
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/error.c:5]: descriptive, but clunky

File: Core/Src/main.c [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+        }
+        
+        if (millis() - last_message_time > MAX_BUS_DEAD_TIME_ms) {
+            // We've got too long without seeing a valid CAN message (including one of ours)
+            //__asm__ volatile ("reset");
+        }
+        
+
+        //blink blue LED at 1/3 Hz, duty cycle of 1/12
+        if (millis() - last_blink_time > 500) {
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/main.c:191]: ahhhhh magic number put it in a define

File: Core/Src/health_check.c [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#define BATTERY_CURRENT_THRESHOLD 400 // Should be even less when running without OV5640
+
+#define BATT_VOLT_SCALAR_TIMES 17 // ADC / 4096 * 2.8V * 5 * 1000 mV/V = ADC * 3.417
+#define BATT_VOLT_SCALAR_DIV 5
+#define BATT_VOLT_LOW_THRESHOLD 8000
+#define BATT_VOLT_HIGH_THRESHOLD 13000 // ADC will saturate by then
+
+bool check_bus_current_error(void) {
+    HAL_ADC_Start(&hadc1);
+    HAL_ADC_PollForConversion(&hadc1, 1000);
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/health_check.c]: Are you sure you want to poll for a second here? If you actually block your while loop for a second I think you will block your "last seen CAN message" thing also for a second and reset the board? Unless thats the intended function?

File: Core/Src/main.c (hunk 3 above)
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/main.c:191]: ahhhhh magic number put it in a define

File: Core/Src/main.c [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+    int dest_id = get_reset_board_id(message);
+
+    //handle a "LED_ON" or "LED_OFF" message
+    // Declare this outside of switch statement to prevent errors
+    int cmd_type = -1;
+    switch (get_message_type(message)) {
+        case MSG_GENERAL_CMD:
+            cmd_type = get_general_cmd_type(message);
+            if (cmd_type == BUS_DOWN_WARNING) {
+                logger_off = 40;
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/main.c:92]: Why does this "stop logging after bus down warning" mechanism exist? Like what problem is this intended to solve?

File: Core/Src/platform.c [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#include "platform.h"
+
+uint32_t millis(void) {
+    return HAL_GetTick();
+}
+
+uint32_t micros(void) {
+    return 0;
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/platform.c:8]: uhhhhhhhh...
`#warning` maybe?

File: Core/Src/platform.c [hunk 7]
```diff
@@ -0,0 +1,9 @@
+#include "platform.h"
+
+uint32_t millis(void) {
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/platform.c:3]: really curious how you got this to compile. Maybe fixed with one of the canlib update and we just need to try again on processor

File: Core/Src/main.c [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+                LED_RED_OFF();
+            }
+            last_blink_time = millis();
+        }
+
+        //give status update
+        if (millis() - last_board_status_msg > 500) {
+            can_msg_t board_stat_msg;
+            bool status_ok = true; //t
+            status_ok = status_ok & !check_bus_current_error();
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/main.c:205]: &=

File: Core/Src/main.c [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+			status_ok = status_ok & !check_bus_voltage_error();
+            // for now just always pretend everything is ok
```
Comment (Joe-Joe-Joe-Joe) [inline Core/Src/main.c]: Is this comment still accurate?

**Review by Joe-Joe-Joe-Joe — COMMENTED** [review]
How thoroughly has it been tested thus far?

**Comment by Joe-Joe-Joe-Joe** [comment]
I also HIGHLY recommend adding this build workflow
https://github.com/waterloo-rocketry/cansw_processor_stm/blob/master/.github/workflows/build.yml
simple sanity check so you aren't inadvertently breaking master

File: COPYING
Comment (JasonBrave) [reply COPYING:1]: License

File: Core/Src/platform.c (hunk 7 above)
Comment (JasonBrave) [reply Core/Src/platform.c:3]: Build file filter

File: Core/Src/main.c (hunk 5 above)
Comment (JasonBrave) [reply Core/Src/main.c:92]: flush buffer, also no bus down warning

**Pull request merged**